### DISCUSSION TITLE
Types/Name autocomplete in custom input components.

### DISCUSSION THREAD

**Discussion by ghost** [react-hook-form/react-hook-form#5536] [Q&A]
Currently, when I use register method on simple input components I have the name autocomplete in the first argument of register method.
Is there any example of cooking correct types/generics for custom input components to keep the autocomplete on the "name" prop?
(custom inputs powered by useController inside)

Thanks in advance

**Comment by bluebill1049** [comment]
can you provide a codesandbox?

**Comment by thanh-nguyen-95** [comment]
If I understand this correctly, you should see code suggestion for `name` right after passing in `control` prop/option (given you have already typed `useForm` or `useFormContext`)

An example:

```js
const { control } = useForm<{ foo: string }>()
const controllerProps = useController({ control, name: /* 'foo' should show up here as a code suggestion in your IDE */ })
```

**Comment by ghost** [comment]
You can use https://codesandbox.io/s/usecontroller-0o8px as an example, I'd like to see suggestion when typing in a "name" prop on line 43
But not sure that that is possible

**Reply by thanh-nguyen-95** [reply]
![image](https://user-images.githubusercontent.com/44762180/121344384-e8c7b900-c94d-11eb-98db-9d3a7a43a983.png)

It's working for me. Does it not show up for you? Try triggering code autocomplete manually by pressing `ctrl + spacebar` or `cmd + spacebar` if you're using macOs.

**Reply by ghost** [reply]
Yes, that works here because we have the Input component in the same file and we can provide the FormValues type directly to the generic type...

But in real project all your components will be in different file and they will be imported in the form component. 
So I'm trying to find a way how to tell my CustomInput component which name it could consume in this particular form

**Reply by thanh-nguyen-95** [reply]
You can try the generic approach like so:

```tsx
function Input<T extends Record<string, any>>(props: UseControllerProps<T>) {
  /* implement your UI here */
}

// usage
<Input<FormValues> control={control} name="/* suggestion should show up here */" />
```

It's not an optimal solution but I can't think of any other way at the moment.
